docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,459 @@
+  type Proposal struct {
+      // ID is the unique identifier of the proposal.
+      ID uint64 `json:"id"`
+      // Submitter is the address of the proposal submitter.
+      Submitter staking.Address `json:"submitter"`
+      // State is the state of the proposal.
+      State ProposalState `json:"state"`
+      // Deposit is the deposit attached to the proposal.
+      Deposit quantity.Quantity `json:"deposit"`
+
+      // Content is the content of the proposal.
+      Content ProposalContent `json:"content"`
+
+      // CreatedAt is the height at which the proposal was created.
+      CreatedAt int64 `json:"created_at"`
+      // ClosesAt is the height at which the proposal will close and votes will
+      // be tallied.
+      ClosesAt int64 `json:"closes_at"`
+
+      // Results are the final tallied results after the voting period has
+      // ended. Note that they take the
+      Results map[Vote]quantity.Quantity `json:"results,omitempty"`
+      // InvalidVotes is the number of invalid votes after tallying.
+      InvalidVotes uint64 `json:"invalid_votes,omitempty"`
+  }
+  ```
+
+- **List of active proposals (`0x82`)**
+
+  Each active proposal (one that has not yet closed) is stored under a separate
+  storage key with the following key format:
+
+  ```
+  0x82 <closes-at (uint64)> <proposal-id (uint64)>
+  ```
+
+  The value is empty as the proposal ID can be inferred from the key.
+
+- **List of votes (`0x83`)**
+
+  Each vote is stored under a separate storage key with the following key
+  format:
+
+  ```
+  0x83 <proposal-id (uint64)> <voter-address (staking.Address)>
+  ```
+
+  And CBOR-serialized value:
+
+  ```golang
+  // Vote is a governance vote.
+  type Vote uint8
+
+  const (
+      VoteYes     Vote = 1
+      VoteNo      Vote = 2
+      VoteAbstain Vote = 3
+  )
+  ```
+
+- **List of pending upgrades (`0x84`)**
+
+  Each pending upgrade is stored under a separate storage key with the following
+  key format:
+
+  ```
+  0x84 <upgrade-epoch (beacon.EpochTime)> <proposal-id (uint64)>
+  ```
+
+### Genesis Document
+
+The genesis document needs to be updated to include a `governance` field with
+any initial state (see [_State_]) and consensus parameters (see [_Consensus
+Parameters_]) for the governance service.
+
+[_State_]: #state
+[_Consensus Parameters_]: #consensus-parameters
+
+### Transaction Methods
+
+This proposal adds the following transaction methods in the governance module:
+
+#### Submit Proposal
+
+Proposal submission enables a new consensus layer governance proposal to be
+created.
+
+**Method name:**
+
+```
+gov.SubmitProposal
+```
+
+**Body:**
+
+```golang
+// ProposalContent is a consensus layer governance proposal content.
+type ProposalContent struct {
+    Upgrade       *UpgradeProposal       `json:"upgrade,omitempty"`
+    CancelUpgrade *CancelUpgradeProposal `json:"cancel_upgrade,omitempty"`
+}
+
+// UpgradeProposal is an upgrade proposal.
+type UpgradeProposal struct {
+    upgrade.Descriptor
```

Comment:
Is https://github.com/oasisprotocol/oasis-core/issues/2596 a prerequisite for this ADR? Or maybe at least a "good to have", so voters can actually confirm the proposed upgrade identifier (hash) matches a specific source code. (i guess alternatively, although not ideal, we could use something like hash of the source code?)

Reply:
Yes, that's right. So in theory the hash could also be a Git commit or a hash of a manifest containing further information.